internal/common/isolationgroup/service_wrapper.go

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package isolationgroup
+
+import (
+	"context"
+
+	"go.uber.org/yarpc"
+
+	"go.uber.org/cadence/.gen/go/cadence/workflowserviceclient"
+	"go.uber.org/cadence/.gen/go/shared"
+)
+
+const (
+	isolationGroupHeader = "cadence-client-zone"
+)
+
+type workflowServiceIsolationGroupWrapper struct {
+	service                  workflowserviceclient.Interface
+	isolationGroupIdentifier string
+}
+
+// NewWorkflowServiceWrapper creates
+func NewWorkflowServiceWrapper(service workflowserviceclient.Interface, isolationGroup string) workflowserviceclient.Interface {
+	return &workflowServiceIsolationGroupWrapper{
+		service:                  service,
+		isolationGroupIdentifier: isolationGroup,
+	}
+}
+
+func (w *workflowServiceIsolationGroupWrapper) getIsolationGroupIdentifier() yarpc.CallOption {
+	return yarpc.WithHeader(isolationGroupHeader, w.isolationGroupIdentifier)
+}
+
+func (w *workflowServiceIsolationGroupWrapper) DeprecateDomain(ctx context.Context, request *shared.DeprecateDomainRequest, opts ...yarpc.CallOption) error {
+	opts = append(opts, w.getIsolationGroupIdentifier())
+	return w.service.DeprecateDomain(ctx, request, opts...)
+}
+
+func (w *workflowServiceIsolationGroupWrapper) ListDomains(ctx context.Context, request *shared.ListDomainsRequest, opts ...yarpc.CallOption) (*shared.ListDomainsResponse, error) {
+	opts = append(opts, w.getIsolationGroupIdentifier())
```

Comment:
given how regular this pattern is, might be worth adopting gowrap to reduce minor mistakes now and in maintenance.

we already have it in the server, it'd be fairly easy to add here: https://github.com/uber/cadence/blob/e2a2a940030ec62a0749a4f4d8c86100305220c2/client/frontend/interface.go#L32